linkerd/http/retry/src/peek_trailers.rs

```diff
@@ -94,55 +108,103 @@ impl<B: Body> PeekTrailersBody<B> {
+                        // We immediately yielded a result, but it was an error. Alas!
+                        Some(Err(error)) => Self::Unary {
+                            data: Some(Ok(first)),
+                            trailers: Some(Err(error)),
+                        },
+                        // We immediately yielded another frame, but it was a second DATA frame.
+                        // We hold on to each frame, but we cannot wait for the TRAILERS.
+                        Some(Ok(Ok(second))) => Self::Unknown {
+                            first: Some(Ok(first)),
+                            second: Some(Ok(second)),
+                            inner: body.into_inner(),
+                        },
+                        // The body immediately yielded a second TRAILERS frame. Nice!
+                        Some(Ok(Err(trailers))) => Self::Unary {
+                            data: Some(Ok(first)),
+                            trailers: Some(Ok(trailers)),
+                        },
+                    }
+                } else {
+                    // If we are here, the second frame is not yet available. We cannot be sure
+                    // that a second DATA frame is on the way, and we are no longer willing to
+                    // await additional frames. There are no trailers to peek.
+                    Self::Unknown {
+                        first: None,
+                        second: None,
+                        inner: body.into_inner(),
+                    }
+                }
+            }
         };
 
-        if trailers.is_some() {
+        if body.peek_trailers().is_some() {
             tracing::debug!("Buffered trailers frame");
         }
 
-        Self {
-            inner: body,
-            first_data,
-            trailers,
-        }
+        body
     }
 
-    /// Returns a response with an inert [`PeekTrailersBody<B>`].
+    /// Splits a `Frame<T>` into a `Result<T, E>`.
     ///
-    /// This will not peek the inner body's trailers.
-    fn no_trailers(rsp: http::Response<B>) -> http::Response<Self> {
-        rsp.map(|inner| Self {
-            inner,
-            first_data: None,
-            trailers: None,
-        })
+    /// Frames do not expose their inner enums, and instead expose `into_data()` and
+    /// `into_trailers()` methods. This function will return `Ok(data)` if it is given a DATA
+    /// frame, and `Err(trailers)` if it is given a TRAILERS frame.
+    ///
+    /// This is an internal helper to facilitate pattern matching in `read_body(..)`, above.
+    fn split_frame(frame: crate::compat::Frame<B::Data>) -> Result<B::Data, HeaderMap> {
+        frame
+            .into_data()
+            .map_err(|frame| match frame.into_trailers() {
+                Ok(trls) => trls,
+                // Safety: this is not reachable, we called `into_data()` above.
+                Err(_) => unreachable!("into_data() and `into_trailers()` both returned `Err(_)`"),
```

Comment:
i'm unsure how much to worry about the commented out `//Unknown(Box<dyn Frameish>),` variant in the vendored frame code. in other words, might this someday be reachable?

`Result<T, E>` gives us something we can pattern match on, which `Frame<T>` does not for the sake of semantic versioning. if we fear hitting this in the future, we could define a local enum like so...

```rust
enum ParsedFrame<T> {
    Data(T),
    Trailers(HeaderMap),
    Other(Frame<T>),
}
```

...and return that here. what do you think?

Reply:
Or this could be `Option<Either<B::Data, HeaderMap>>`, and then the caller can be adapted so this is handled as an Empty body... Since the API specifically doesn't give us exhaustive matching it's probably best to handle it cleanly.

Reply:
agreed. this is addressed in 501b3cf0